File: Core/Hooks/DVTOnboardingTutorialController.m
#import <objc/runtime.h>
#import "DVTDeviceKit.h"
#import "IDEAVRPlatformSupportCore-Swift.h"

@implementation DVTOnboardingTutorialController (Hooks)

+ (void)load {
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		Method originalMethod = class_getInstanceMethod([self class], @selector(pageController:viewControllerForIdentifier:));
		Method swizzledMethod = class_getInstanceMethod([self class], @selector(_pageController:viewControllerForIdentifier:));
		method_exchangeImplementations(originalMethod, swizzledMethod);
	});
}

- (DVTOnboardingPage*)_pageController:(DVTPageController*)v1 viewControllerForIdentifier:(NSString*)v2 {
	if ([v2 isEqualToString:@"setup-device"] && [[self selectedDeviceForSetup] isKindOfClass:[DVTAVRDevice class]]) {
		DVTOnboardingPage* page = [[DVTOnboardingTutorialPairDevice alloc] init];
		[page setTutorialController:self];
		return page;
	} else {
		return [self _pageController:v1 viewControllerForIdentifier:v2];
	}
}

@end

File: Core/DVTOnboardingTutorialController.swift
//
//  DVTOnboardingTutorialController.swift
//  Core
//
//  Created by Levi Dhuyvetter on 10/12/2020.
//

import Foundation

extension DVTOnboardingTutorialController {
	class func avrSwizzle() {
		DVTOnboardingTutorialSuccess.avrSwizzle()
		let originalClass = DVTOnboardingTutorialController.self
		let originalMethod = class_getInstanceMethod(originalClass, #selector(DVTOnboardingTutorialController.pageController(_:viewControllerForIdentifier:)))
		let newMethod = class_getInstanceMethod(originalClass, #selector(DVTOnboardingTutorialController.avrPageController(_:viewControllerForIdentifier:)))
		method_exchangeImplementations(originalMethod!, newMethod!)
	}
	
	@objc dynamic func avrPageController(_ v1: DVTPageController!, viewControllerForIdentifier identifier: String!) -> DVTOnboardingPage! {
		if identifier == "setup-device", let _ = self.selectedDeviceForSetup as? DVTAVRDevice {
			let page = DVTOnboardingTutorialPairDevice()
			page.tutorialController = self
			
			return page
		} else {
			//Calling itself but is actually old implementation when methods are swizzled.
			return self.avrPageController(v1, viewControllerForIdentifier: identifier)
		}
	}
}

extension DVTOnboardingTutorialSuccess {
	class func avrSwizzle() {
		let originalClass = DVTOnboardingTutorialSuccess.self
		let originalMethod = class_getInstanceMethod(originalClass, #selector(DVTOnboardingTutorialSuccess.viewDidAppear))
		let newMethod = class_getInstanceMethod(originalClass, #selector(DVTOnboardingTutorialSuccess.avrViewDidAppear))
		method_exchangeImplementations(originalMethod!, newMethod!)
	}
	
	@objc dynamic func avrViewDidAppear() {
		(self.tutorialController.selectedDeviceForSetup as? DVTAVRDevice)?.paired = true
		
		//Calling itself but is actually old implementation when methods are swizzled.
		self.avrViewDidAppear()
	}
}
